notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,304 @@
+  """
+  def prepare_training_data(images, batch_size) do
+    batched_images = images |> Nx.to_batched(batch_size)
+    Stream.zip(batched_images, batched_images)
+  end
+end
+```
+
+```elixir
+train_images = Utils.preprocess_data(Scidata.FashionMNIST.download())
+test_images = Utils.preprocess_data(Scidata.FashionMNIST.download_test())
+
+Kino.render(train_images[[images: 0]] |> Utils.image_to_kino())
+Kino.render(test_images[[images: 0]] |> Utils.image_to_kino())
+
+:ok
+```
+
+Our simple autoencoder model. We won't be using a denoising autoencoder here.
+
+Note that we're giving each of the layers a name - the reason for this will be apparent later.
+
+```elixir
+model =
+  Axon.input("image", shape: {nil, 1, 28, 28})
+  # This is now 28*28*1 = 784
+  |> Axon.flatten()
+  # The encoder
+  |> Axon.dense(256, activation: :relu, name: "encoder_layer_1")
+  |> Axon.dense(128, activation: :relu, name: "encoder_layer_2")
+  |> Axon.dense(64, activation: :relu, name: "encoder_layer_3")
+  # Bottleneck layer
+  |> Axon.dense(10, activation: :relu, name: "bottleneck_layer")
+  # The decoder
+  |> Axon.dense(64, activation: :relu, name: "decoder_layer_1")
+  |> Axon.dense(128, activation: :relu, name: "decoder_layer_2")
+  |> Axon.dense(256, activation: :relu, name: "decoder_layer_3")
+  |> Axon.dense(784, activation: :sigmoid, name: "decoder_layer_4")
+  # Turn it back into a 28x28 single channel image
+  |> Axon.reshape({1, 28, 28})
+
+# We can use Axon.Display to show us what each of the layers would look like
+# assuming we send in a batch of 4 images
+Axon.Display.as_table(model, Nx.template({4, 1, 28, 28}, :f32)) |> IO.puts()
+```
+
+```elixir
+batch_size = 128
+
+train_data = Utils.prepare_training_data(train_images, 128)
+test_data = Utils.prepare_training_data(test_images, 128)
+
+{input_batch, target_batch} = Enum.at(train_data, 0)
+Kino.render(input_batch[[images: 0]] |> Utils.image_to_kino())
+Kino.render(target_batch[[images: 0]] |> Utils.image_to_kino())
+
+:ok
+```
+
+When training, it can be useful to stop execution early - either when you see it's failing and you don't want to waste time waiting for the remaining epochs to finish, or if it's good enough and you want to start experimenting with it.
+
+The `livebook_early_stop/1` function below is a handy handler to give us a `Kino.Control.button` that will stop the training loop when clicked.
+
+```elixir
+defmodule LivebookTraining do
+  def livebook_early_stop(loop) do
+    frame = Kino.Frame.new() |> Kino.render()
+    stop_button = Kino.Control.button("stop")
+    Kino.Frame.render(frame, stop_button)
+
+    {:ok, button_agent} = Agent.start_link(fn -> nil end)
+
+    Task.start_link(fn ->
+      Kino.Control.subscribe(stop_button, :stop)
+
+      receive do
+        {:stop, _info} ->
+          Kino.Control.unsubscribe(stop_button)
+          Agent.update(button_agent, fn _ -> :stop end)
+      end
+    end)
```

Comment:
```suggestion
    Kino.Control.subscribe(stop_button, fn _event ->
      Agent.update(button_agent, fn _ -> :stop end)
    end)
```

This should do the job :D It runs asynchronously, and since we remove the button from the frame, the handler is unsubscribed automatically.

---

Sidenote: we will have more flexible APIs for dealing with events/streams in the next kino release. To asynchronously consume just one click we would do:

```elixir
stop_button
|> Kino.Control.stream()
|> Stream.take(1)
|> Kino.listen(fn _event ->
  Agent.update(button_agent, fn _ -> :stop end)
end)
```